File: next/Session/ScriptViews.fs
namespace Wanxiangshu.Next.Session

open Wanxiangshu.Next.Kernel
open Wanxiangshu.Next.Kernel.Identity
open Wanxiangshu.Next.Kernel.Outcome
open Wanxiangshu.Next.Journal

module SessionScriptViews =

    let getTodo (gateway: IGateway) (sessionId: SessionId) () : TodoView =
        match Map.tryFind sessionId gateway.ProjectionSet.SessionProjections with
        | Some proj ->
            match proj.Todos with
            | Some snap ->
                { Unfinished = not (List.isEmpty snap.Items)
                  ProgressStamp = proj.Version }
            | None ->
                { Unfinished = false
                  ProgressStamp = proj.Version }
        | None ->
            { Unfinished = false
              ProgressStamp = 0L }

    let getReview (gateway: IGateway) (sessionId: SessionId) (config: SessionScriptConfig) () : ReviewView =
        match Map.tryFind sessionId gateway.ProjectionSet.SessionProjections with
        | Some proj ->
            let req =
                match proj.LastReview with
                | Some Fact.ReviewVerdict.Passed -> false
                | _ -> true
            { Required = req
              Round = 0
              MaxRound = config.MaxInvalidRetries
              Verdict = proj.LastReview }
        | None ->
            { Required = true
              Round = 0
              MaxRound = config.MaxInvalidRetries
              Verdict = None }

    let getProgressStamp (gateway: IGateway) (sessionId: SessionId) () : int64 =
        match Map.tryFind sessionId gateway.ProjectionSet.SessionProjections with
        | Some proj -> proj.Version
        | None -> 0L

File: .wanxiangshu/runtimes/Session/ScriptViews.js

import { tryFind } from "../fable_modules/fable-library-js.5.6.0/Map.js";
import { ReviewView, TodoView } from "./ScriptTypes.js";
import { isEmpty } from "../fable_modules/fable-library-js.5.6.0/List.js";

export function getTodo(gateway, sessionId, unitVar) {
    const matchValue = tryFind(sessionId, gateway.ProjectionSet.SessionProjections);
    if (matchValue == null) {
        return new TodoView(false, 0n);
    }
    else {
        const proj = matchValue;
        const matchValue_1 = proj.Todos;
        if (matchValue_1 == null) {
            return new TodoView(false, proj.Version);
        }
        else {
            return new TodoView(!isEmpty(matchValue_1.Items), proj.Version);
        }
    }
}

export function getReview(gateway, sessionId, config, unitVar) {
    let matchValue_1;
    const matchValue = tryFind(sessionId, gateway.ProjectionSet.SessionProjections);
    if (matchValue == null) {
        return new ReviewView(true, 0, config.MaxInvalidRetries, undefined);
    }
    else {
        const proj = matchValue;
        return new ReviewView((matchValue_1 = proj.LastReview, (matchValue_1 != null) ? (!(matchValue_1.tag === 0)) : true), 0, config.MaxInvalidRetries, proj.LastReview);
    }
}

export function getProgressStamp(gateway, sessionId, unitVar) {
    const matchValue = tryFind(sessionId, gateway.ProjectionSet.SessionProjections);
    if (matchValue == null) {
        return 0n;
    }
    else {
        return matchValue.Version;
    }
}

